File: 1.Two Sum/solution.py
class Solution:
    def twoSum(self, num,target):
        
        # declare hash map for storing: complement -> key
        hash_map = {}
    
        # create and check for complement
        for index1,value in enumerate(num):

            # check if complement in hash map
            # then we can return
            complement = target-value

            if complement in hash_map:          
                index2 = hash_map[complement]
                if index1 != index2:
                    return index1, index2
            else:
                hash_map[value] = index1

File: 1.Two Sum/solution.java
/*
 * The best solution for 2 sum problem in java can do it in 1 pass
 * While we iterate and insert elements into the table , we can 
 * check whether the complement exists or not.
 */

import java.util.Map;
import java.util.HashMap;

/**
 * Function calculates 2 sum
 * returns indexes of numbers which
 * evaluate to the target sum
 */
public int[] twoSum(int[] nums, int target) {
	
	// declare and initialize map
	Map<Integer, Integer> map = new HashMap<>();
	
	// only 1 pass needed this time
	for (int i = 0; i < nums.length; i++) {
		
		int complement = target - nums[i];
		
		// remember we are ommiting the statement
		// which checked the existence of same number
		// because we haven't inserted the number yet
		if (map.containsKey(complement)) {
			return new int[] { map.get(complement), i };
		}
		else {
			map.put(nums[i],i);
		}
	}
	return new int[] {-1,-1};
}
